File: TopCoder/src/main/java/com/leetcode/String/ZigZagConversion.java
package com.leetcode.String;

//https://leetcode.com/problems/zigzag-conversion/description/
public class ZigZagConversion {

    public String zigzagprint (String str, int rowNumber) {
        if (str.length() <= rowNumber) {
            return str;
        }
        int startIndex = 0;
        int originalGap = 2 * rowNumber - 2;
        if (originalGap == 0) {
            return str;
        }
        int nextIndex = startIndex + originalGap;
        int currentOddGap;
        int currentEvenGap;
        StringBuilder stringBuilder = new StringBuilder();
        for ( int i = 0 ; i < rowNumber ; i++) {
            if ( i == 0 || i == rowNumber -1) {
                for (int j = startIndex; j < str.length(); j = j + originalGap) {
                    stringBuilder.append(str.charAt(j));
                }
            } else {
                int currentIndex = startIndex;
                int nextOriginalNext = startIndex + originalGap;
                int iterationNum = 0;
                currentEvenGap = (nextIndex - startIndex);
                currentOddGap = (nextOriginalNext - nextIndex);
                while (currentIndex < str.length()) {
                    stringBuilder.append(str.charAt(currentIndex));
                    if (iterationNum % 2 == 0) {
                        currentIndex += currentEvenGap;
                    } else {
                        currentIndex += currentOddGap;
                    }
                    iterationNum++;
                }
            }
            startIndex++;
            nextIndex--;
        }
        return stringBuilder.toString();
    }

    public static void main(String[] args) {
        String input = "AB";
        int rowNumber = 1;
        String result = new ZigZagConversion().zigzagprint(input, rowNumber);
        System.out.println(result);
    }
}

File: TopCoder/python/arr/ZigZagConversion.py
#https://leetcode.com/problems/zigzag-conversion/description/

def zigzagprint (s, numRows):
    if len(s) <= numRows:
        return s
    startIndex = 0
    originalGap = 2 * numRows - 2
    if originalGap == 0:
        return s
    nextEndIndex = startIndex + originalGap
    retVal = ""
    for i in range(0, numRows):
        j = startIndex
        if i == 0 or i == numRows - 1:
            while j < len(s):
                retVal += s[j]
                j = j + originalGap
        else:
            nextEvenGap = (nextEndIndex - startIndex)
            nextOddGap = (startIndex + originalGap) - nextEndIndex
            #retVal += str[j]
            iteration = 0
            while j < len(s):
                retVal += s[j]
                if iteration % 2 == 0 :
                    j = j + nextEvenGap
                else:
                    j = j + nextOddGap
                iteration = iteration +1

        startIndex=startIndex+1
        nextEndIndex=nextEndIndex-1

    return retVal


str1 = "PAYPALISHIRING"
row = 4
retVal = zigzagprint(str1, row)
print(retVal)